web/packages/teleterm/src/ui/TopBar/Connections/connectionsContext.tsx

```diff
@@ -0,0 +1,67 @@
+/**
+ * Teleport
+ * Copyright (C) 2024 Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import {
+  FC,
+  PropsWithChildren,
+  createContext,
+  useCallback,
+  useContext,
+  useState,
+} from 'react';
+
+/**
+ * ConnectionsContext allows other parts of the app to control the connection list.
+ */
+export type ConnectionsContext = {
+  isOpen: boolean;
+  close: () => void;
+  toggle: () => void;
+};
+
+export const ConnectionsContext = createContext<ConnectionsContext>(null);
+
+export const ConnectionsContextProvider: FC<PropsWithChildren> = props => {
+  const [isOpen, setIsOpen] = useState(false);
+
+  const toggle = useCallback(() => {
+    setIsOpen(wasOpen => !wasOpen);
+  }, []);
+
+  const close = useCallback(() => {
+    setIsOpen(false);
+  }, []);
```

Comment:
i rarely use `useCallback` so i tend to forget how it helps 😅 , like i wanted to understand why we used it here? will it prevent like a re-render somewhere

Reply:
In the case of `toggle`, it's going to prevent unnecessarily firing a hook on each render of `Connections`:

https://github.com/gravitational/teleport/blob/a76b1e8f9b0fb12544e3ae8e4226188c755a8645/web/packages/teleterm/src/ui/TopBar/Connections/Connections.tsx#L41-L48

https://github.com/gravitational/teleport/blob/a76b1e8f9b0fb12544e3ae8e4226188c755a8645/web/packages/teleterm/src/ui/services/keyboardShortcuts/useKeyboardShortcuts.ts#L28-L40

When defining a function that is passed from a parent to a child, `useCallback` often doesn't matter much since usually a child is going to get re-rendered when its parent re-renders anyway. But when writing a context, I often don't know upfront if functions returned from the context are going to be used as effect dependencies. In that case, it's just feels safer for me to start with `useCallback`.

Hopefully this will be rendered obsolete by React Compiler aka React Forget.

* https://react.dev/blog/2024/02/15/react-labs-what-we-have-been-working-on-february-2024#react-compiler
* https://www.youtube.com/watch?v=qOQClO3g8-Y